src/main/scala/essential_scala/Ex4Traits.scala

```diff
@@ -86,24 +88,64 @@ object Ex4Traits extends App {
   // 4.6.3.2 The Forest of Trees
   // pattern matching
   sealed trait Tree {}
-  case class Node(left: Tree, right: Tree) extends Tree
-  case class Leaf(element: Int) extends Tree
-  object Math {
-    def prettyPrint(t: Tree, indent: String): Unit = t match {
-      case (t: Node) =>
-        prettyPrint(t.left, indent + " ")
-        prettyPrint(t.right, indent + " ")
-        println(indent + ".")
-      case (t: Leaf) => println(indent + t.element)
+  final case class Node(left: Tree, right: Tree) extends Tree
+  final case class Leaf(element: Int) extends Tree
+
+  object Tree {
+    def mkSpace(i: Int) = Array.fill(i) { " " }.mkString("")
+    def prettyPrint(t: Tree): String = {
+      @tailrec
+      def inner(l: List[(Tree, Int)], indent: Int, acc: String): String =
+        l match {
+          case Nil => acc
+          case (Leaf(v), ind) :: ls =>
+            inner(ls.map(el => (el._1, el._2)),
```

Comment:
What is `ls.map(el => (el._1, el._2))` for? Isn't it equal to `ls` in this case?